### ISSUE TITLE
ubiattach error 22 on TS-932PX

### ISSUE THREAD

**Issue by coreyshuman** [OneCDOnly/create-autorun#12]
When running the create-autorun script, the script fails to mount the config partition:

```
[~] # curl -skL https://git.io/create-autorun | sudo bash
create-autorun.sh 230207

info: NAS model: TS-932PX
info: QTS version: 5.0.1 #20230112
info: default volume: /share/CACHEDEV1_DATA
info: autorun partition should be: /dev/mtd6
skip: unable to ubiattach 'ubiattach: error!: cannot attach mtd6
           error 22 (Invalid argument)'
info: will try as ext4 instead
fail: Unable to mount ext4 autorun partition: /dev/mtd6 'mount: special device /dev/mmcblk0p7 does not exist'
```

Running the command `cat /proc/mtd` it would appear that `mtd6` is the correct device for the config region:

```
[~] # cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00150000 00001000 "loader"
mtd1: 00060000 00001000 "env"
mtd2: 02000000 00020000 "boot1_kernel"
mtd3: 0d800000 00020000 "boot1_rootfs2"
mtd4: 02000000 00020000 "boot2_kernel"
mtd5: 0d800000 00020000 "boot2_rootfs2"
mtd6: 00f00000 00020000 "config"
```

I tried to manually run `ubiattach` with a few different argument syntaxes to no avail:

```
[~] # sudo /sbin/ubidetach /dev/ubi_ctrl -m 6
ubidetach: error!: cannot detach mtd6
           error 19 (No such device)
[~] # sudo /sbin/ubiattach /dev/ubi_ctrl -m 6
ubiattach: error!: cannot attach mtd6
           error 22 (Invalid argument)
[~] # ubiattach -m 6 -d 2
ubiattach: error!: cannot attach mtd6
           error 22 (Invalid argument)
[~] # sudo /sbin/ubiattach -p /dev/mtd6
ubiattach: error!: cannot attach "/dev/mtd6"
           error 22 (Invalid argument)
[~] # sudo /sbin/ubiattach -p /dev/mtd6ro
ubiattach: error!: cannot attach "/dev/mtd6ro"
           error 22 (Invalid argument)
```

Out of desperation I tried running `ubiattach` on `mtd5` and that was successful:

```
[~] # sudo /sbin/ubiattach -p /dev/mtd5
UBI device number 0, total 1728 LEBs (219414528 bytes, 209.2 MiB), available 0 LEBs (0 bytes), LEB size 126976 bytes (124.0 KiB)
```

I'm not sure what the next step is to troubleshoot this. Any suggestions are appreciated.

**Comment by OneCDOnly** [comment]
Good work. This looks like another NAS model that requires the 5th partition instead of the 6th. There's a list of model exceptions to specify these. 

I can add your model NAS to this list. What is the result of this command?
```
getcfg System Model
```

**Comment by coreyshuman** [comment]
```
[~] # getcfg System Model
TS-X32
```

**Comment by coreyshuman** [comment]
Based on the `platform.conf` file, the kernel thinks it should be `mtd6`:

```
[~] # cat /etc/platform.conf
Platform = ARM_AL2
SUID = ...

[DIMM 1]
valid = 1
channel_id = 1
slot_id = 1
dimm_id = 1
slot_addr = 0x56
type = 12
speed = 9
module_type = 3
ecc = 0
size = 16384
manufaturer = Crucial Technology

[FIRMWARE STORAGE]
DEVICE_NODE = /dev/mtd
FS_TYPE = ubifs
UPDATE_TMP_PARTITION = 5
FS_ACTIVE_PARTITION = 3
FS_RECOVERY_PID = -1

[CONFIG STORAGE]
DEVICE_NODE = /dev/mtd
FS_TYPE = ubifs
FS_ACTIVE_PARTITION = 6

[MISC]
DISPLAY_NAME = TS-932PX
```

And the built-in `/etc/init.d/init_disk.sh mount_flash_config` command is failing as well. I wonder if I just have a corrupted config partition.

**Comment by coreyshuman** [comment]
Further confirming my suspicions that the config might be corrupted, I found a `check_config` command which also fails, and attempts but fails to recover the partition.

```
[~] # /etc/init.d/init_disk.sh check_config
Check config files...start
Mount /dev/mtd6 failed...try to recover it...
/opt/nasconfig_fs.img.tgz is found.
```

**Comment by OneCDOnly** [comment]
I've just updated the script with an exception for your model. 

Can you please try running it again? 
```
curl -skL https://git.io/create-autorun | sudo bash
```

**Comment by coreyshuman** [comment]
It is still using `mtd6` for some reason:

```
[~] # getcfg System Model
TS-X32
[~] # curl -skL https://git.io/create-autorun | sudo bash
create-autorun.sh 230215

info: NAS model: TS-932PX
info: QTS version: 5.0.1 #20230112
info: default volume: /share/CACHEDEV1_DATA
info: autorun partition should be: /dev/mtd6
skip: unable to ubiattach 'ubiattach: error!: cannot attach mtd6
           error 22 (Invalid argument)'
info: will try as ext4 instead
fail: Unable to mount ext4 autorun partition: /dev/mtd6 'mount: special device /dev/mmcblk0p7 does not exist'
```

**Comment by OneCDOnly** [comment]
Oops, my apologies. I added the exception to the wrong section of code. I've just pushed another update to GitHub. 

Can you please try again?

**Comment by coreyshuman** [comment]
It still tries to mount the wrong partition since the `ubiattach` command ultimately uses the `$NAS_AUTORUN_PART` variable:

```
[~] # /sbin/getcfg 'CONFIG STORAGE' FS_ACTIVE_PARTITION -f /etc/platform.conf
6
[~] # curl -skL https://git.io/create-autorun | sudo bash
create-autorun.sh 230216

info: NAS model: TS-932PX
info: QTS version: 5.0.1 #20230112
info: default volume: /share/CACHEDEV1_DATA
info: autorun partition should be: /dev/mtd5
skip: unable to ubiattach 'ubiattach: error!: cannot attach mtd6
           error 22 (Invalid argument)'
info: will try as ext4 instead
fail: Unable to mount ext4 autorun partition: /dev/mtd5 'mount: special device /dev/mmcblk0p7 does not exist'
```

I pulled down the script and hardcoded `NAS_AUTORUN_PART = 5` to test, which allowed it to get past attach but ran into a new issue where it does not recognize the `ubifs` filesystem.

```
[~] # ./create-autorun.sh
create-autorun.sh 230216

info: NAS model: TS-932PX
info: QTS version: 5.0.1 #20230112
info: default volume: /share/CACHEDEV1_DATA
info: autorun partition should be: /dev/mtd5
done: ubiattached autorun partition: 5
fail: Unable to mount ubifs autorun partition: /dev/mtd5 'mount: unknown filesystem type 'ubifs''
```

I'll continue to look into this.

**Comment by coreyshuman** [comment]
Tried `ext4` but no luck.

```
[~] # mount -t ext4 ubi2:config /dev/mtd5
mount: special device ubi2:config does not exist
```

And we already know `/dev/mmcblk0p7` doesn't exist either.

**Comment by OneCDOnly** [comment]
If you're able to figure this out, please let me know what is required. 

I'll revert that last change, as it doesn't workaround the issue. 

Cheers.

**Comment by OneCDOnly** [comment]
Hi Corey, I'll close this issue for-now. If you can devise a solution to the described problem, please feel free to reopen this one. 

Cheers!

**Comment by alirezafarnoosh** [comment]
I have ts832-px and script runs well on my nas last week or two . I can provide info in case u guys need it

**Comment by coreyshuman** [comment]
After 3 and a half years later, I was FINALLY able to troubleshoot this and get `autostart.sh` working on the TS-932PX with help from Claude (I wasn't about to recognize the difference between UBI and ext2 raw header bytes). I'm sharing the root cause and recovery steps in case anyone comes across this in the future.

## Summary

On the TS-932PX, `/opt/nasconfig_fs.img.tgz` contains an image in the wrong format for the recovery procedure used by `/etc/init.d/init_disk.sh check_config`.

The recovery procedure writes an ext2 filesystem image directly to raw NAND using `nandwrite`, but this device requires a ubinized UBI image containing a layout volume. As a result, recovery fails and rewrites the invalid image each time it runs. This leaves the partition in a state where `ubiattach` consistently fails with error 22.

The solution is to erase mtd6 and allow UBI to initialize the partition directly.

## Symptoms

* `ubiattach` on mtd6 fails with `error 22 (Invalid argument)`.

* `dmesg` reports:

  ```text
  ubi2: scanning is finished
  ubi2 error: ubi_read_volume_table: the layout volume was not found
  ubi2 error: ubi_attach_mtd_dev: failed to attach mtd6, error -22
  ```

* `check_config` reports the following on every run, but the recovery does not persist:

  ```text
  Mount /dev/mtd6 failed...try to recover it...
  ```

* `autorun.sh` cannot be created because scripts that depend on the flash configuration volume fail at the mount step.

* The NAS otherwise operates normally. QTS keeps its active configuration on the data volume; the flash configuration serves as a bootstrap copy.

## Root cause

`check_config` in `/etc/init.d/init_disk.sh` has separate recovery procedures for two types of configuration storage:

* On ext2-config models, it extracts a filesystem image directly onto a block device:

  ```sh
  tar xzf $NASCONFIG_FS -O > $DEV_NAS_CONFIG
  ```

* On UBIFS-config models, including the TS-932PX, it erases the NAND partition and writes the extracted image with `nandwrite`:

  ```sh
  flash_erase ...
  tar xzf $NASCONFIG_FS -O | nandwrite -p /dev/mtd6
  ```

The second procedure requires the archive to contain a raw UBI container image created with `ubinize`, including the UBI layout volume. On the TS-932PX, however, the archive contains an ext2 filesystem image.

This can be verified with:

```sh
tar xzf /opt/nasconfig_fs.img.tgz -O > /tmp/nasconfig.img
hexdump -C /tmp/nasconfig.img | head -4
```

A UBI image begins with the bytes `55 42 49 23`, representing `UBI#`.

The supplied image instead contains 1,024 zero bytes followed by an ext2 superblock at offset `0x400`, with the ext2 magic value `53 ef` at offset `0x438`. It can also be mounted as an ext2 loop device and contains only `lost+found`.

The image appears to be an empty formatted filesystem intended for QTS to populate at runtime. Writing that ext2 image directly to NAND does not create a valid UBI layout volume. The resulting sequence is:

1. `nandwrite` writes the ext2 image to mtd6.
2. `ubiattach` cannot find a UBI layout volume.
3. Attachment fails with error 22.
4. The next recovery attempt writes the same incompatible image again.

The recovery commands redirect their output to `/dev/null`, and the resulting partition is not validated, so the failure is not reported clearly during startup.

## Important safety note

`check_config` does not first verify that the configuration volume is unmounted. If it is run while the volume is mounted, its own mount attempt fails with `EBUSY`. It then enters the recovery branch and erases and rewrites the partition beneath the mounted filesystem.

Run `check_config` only after the configuration volume has been fully unmounted and detached.

## Repair procedure

> Warning: Erasing mtd6 is appropriate only when the partition is already affected and cannot be mounted. If mtd6 currently mounts successfully, do not use this procedure.

No replacement factory image is required. UBI can initialize an erased NAND partition directly, after which QTS can populate the empty configuration volume.

```sh
# Ensure the configuration volume is not mounted or attached
umount /tmp/nasconfig_tmp 2>/dev/null
/sbin/ubidetach -m 6 2>/dev/null

# Erase the partition and create a new UBI configuration volume
/sbin/flash_erase /dev/mtd6 0 0
/sbin/ubiattach -m 6 -d 2
/sbin/ubimkvol /dev/ubi2 -N config -m
/sbin/ubidetach -m 6

# Allow QTS to populate the configuration files.
# This must be run while the partition is detached.
/etc/init.d/init_disk.sh check_config
```

During `ubiattach`, the output should indicate that an empty MTD device was detected and initialized.

`check_config` should complete with output similar to:

```text
Check config files...start ... end
```

Warnings about a missing `system.map.key` during the first run are expected.

Verify the result with:

```sh
/etc/init.d/init_disk.sh mount_flash_config
mount | grep ubifs
```

Expected results include:

```text
Mount flash config : success
ubi2:config on /tmp/nasconfig_tmp
```

After verification, create `/tmp/nasconfig_tmp/autorun.sh`, make it executable, and then unmount and detach the configuration volume.

Also enable:

**Control Panel → System → Hardware → General → Run user-defined startup processes (autorun.sh)**

The repaired configuration and `autorun.sh` should then persist across reboots.

This procedure has been confirmed on a TS-932PX running QTS 5.2.9 as of August 2026.

## Applicability and limitations

* This procedure applies only to models that use a UBI-backed flash configuration partition. Confirm that `/proc/mtd` contains a `config` partition and that the system reports the UBIFS attachment errors described above. x86 models with a boot DOM and ARM models that use an ext2 configuration partition have different storage layouts and require different procedures.
* The affected firmware still contains the incompatible recovery image. If a future event causes `check_config` to enter its recovery branch again, it may restore the partition to the same invalid state.
* Rebuilding the partition takes approximately 30 seconds. The commands should be retained in case the procedure must be repeated.

**Comment by OneCDOnly** [comment]
Excellent work! 🤓
